### ISSUE TITLE
Feature request: Generalized eigh

### ISSUE THREAD

**Issue by jackd** [google/jax#5461]
[scipy.linalg.eigh](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html) supports solutions to the generalized eigenvalue problem (i.e. when `b is not None`). [jax.scipy.linalg.eigh](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.linalg.eigh.html) claims to support the same interface, but in practice is [NotImplemented](https://github.com/google/jax/blob/master/jax/_src/scipy/linalg.py#L82). I would find this very useful.

```python
import jax
import numpy as np
import scipy.linalg

n = 5
A = np.random.uniform(size=(n, n))
A = A.T @ A
B = np.random.uniform(size=(n, n)
B = B.T @ B

vals, vecs = scipy.linalg.eigh(A, B)  # works
jax_vals, jax_vecs = jax.scipy.linalg.eigh(A, B)
# NotImplementedError("Only the b=None case of eigh is implemented")
```

I imagine part of this would involve extending [this work](https://github.com/google/jax/pull/1099) to cover [sygvd](https://docs.nvidia.com/cuda/cusolver/index.html#sygvd-example1), but I don't really know enough of low-level wrapping to make a contribution.

**Comment by shoyer** [comment]
Yes, this would be nice to have. We would also want to figure out the derivative rule (JVP) for generalized eigenvalue problems.

**Comment by hawkinsp** [comment]
Yes, I think you're correct: adding the primitive is mostly a case of copying the existing `eigh` primitive and extending it to call `sygvd` on CPU and GPU. This should be fairly mechanical, although as always we're pretty busy: contributions are definitely welcome.

Figuring out the gradient might be slightly harder, but it isn't necessary for a first PR.

**Comment by jackd** [comment]
So... having had a stab at the mechanical part of this I realize I'm way out of my depth. I'll try and put together a cpu version modelled after [this PR](https://github.com/google/jax/pull/707/files), but it might not be anytime soon.

On the other hand, I had some fun working through the `jvp` maths and got something that seems to work on the forward pass (at least for real inputs, which is all I'm thinking about initially). I made some notes [here](https://jackd.github.io/posts/generalized-eig-jvp/) and also a [gist](https://gist.github.com/jackd/99e012090a56637b8dd8bb037374900e).

I'm more than happy to test / document the poop out of anything that might arise, or do any general leg-work that doesn't require a good understanding of jax internals.

**Comment by scraed** [comment]
It seems the generalized eigh still not implemented until now. I encountered this issue too and figured out a solution for anyone still need it @jakevdp @hawkinsp  :
```
def generalized_eigh(A, B):
    L = jnp.linalg.cholesky(B)
    L_inv = jnp.linalg.inv(L)
    A_redo = L_inv.dot(A).dot(L_inv.T)
    return jnp.linalg.eigh( A_redo )
```
It may not be the solution with optimal performance, but it works for both CPU and GPU.

**Comment by PabloAMC** [comment]
It seems there is a solution that works here https://gist.github.com/jackd/99e012090a56637b8dd8bb037374900e

**Comment by jackbaker1001** [comment]
I would just like to point out that while @scraed 's answer returns correct eigenvalues, the eigenvectors need to be transformed to have them represent the eigenvectors of the original generalized eigenproblem. I.e, try:

```python
def generalized_eigh(A, B):
    L = np.linalg.cholesky(B)
    L_inv = np.linalg.inv(L)
    C = L_inv @ A @ L_inv.T
    eigenvalues, eigenvectors_transformed = np.linalg.eigh(C)
    eigenvectors_original = L_inv.T @ eigenvectors_transformed
    return eigenvalues, eigenvectors_original
```

**Comment by scraed** [comment]
@jackbaker1001 is right. I forgot to transform the eigen vectors.
